firedrake/assemble.py

```diff
@@ -416,7 +416,9 @@ def visitor(e, *operands):
 
     if tensor:
         update_tensor(result, tensor)
-    return result
+        return tensor
+    else:
+        return result
```

Comment:
That won't work because the tensor is not updated during the DAG traversal, that is precisely why we have this `update_tensor` stage at the end which copies the value of the result into the tensor.